File: sdk/steerplane/loop_detector.py
"""
SteerPlane SDK — Loop Detection Engine

Sliding window pattern detector that identifies repeating action sequences.
Detects patterns like [A, B, A, B, A, B] or [A, A, A, A, A].
"""

from dataclasses import dataclass, field


@dataclass
class LoopDetectionResult:
    """Result of a loop detection check."""
    loop_detected: bool
    pattern: list[str] = field(default_factory=list)
    repetitions: int = 0
    window_size: int = 0


class LoopDetector:
    """
    Sliding window loop detector.
    
    Monitors action history and detects repeating sequences.
    When a loop is detected, the run should be terminated.
    """

    def __init__(self, window_size: int = 8, min_repetitions: int = 2):
        """
        Args:
            window_size: Number of recent actions to analyze.
            min_repetitions: Minimum repetitions to confirm a loop.
        """
        self.window_size = window_size
        self.min_repetitions = min_repetitions
        self.action_history: list[str] = []

    def record_action(self, action: str) -> LoopDetectionResult:
        """
        Record an action and check for loops.
        
        Args:
            action: The action name (e.g., 'search_web', 'call_api').
            
        Returns:
            LoopDetectionResult with detection status and pattern info.
        """
        self.action_history.append(action)
        return self.check()

    def check(self) -> LoopDetectionResult:
        """Check the current action history for loops."""
        if len(self.action_history) < self.window_size:
            return LoopDetectionResult(loop_detected=False)

        return detect_loop(
            self.action_history,
            window_size=self.window_size,
            min_repetitions=self.min_repetitions
        )

    def reset(self):
        """Clear the action history."""
        self.action_history.clear()


def detect_loop(
    history: list[str],
    window_size: int = 8,
    min_repetitions: int = 2
) -> LoopDetectionResult:
    """
    Detect repeating action sequences using a sliding window.
    
    Algorithm:
    1. Take the last `window_size` actions
    2. For each possible pattern length (1 to window/2):
       - Extract the candidate pattern
       - Count how many times it repeats consecutively
       - If repetitions >= min_repetitions, it's a loop
    
    Examples:
        ['search', 'search', 'search', 'search']  → Loop (pattern: ['search'])
        ['A', 'B', 'A', 'B', 'A', 'B']            → Loop (pattern: ['A', 'B'])
        ['A', 'B', 'C', 'D', 'E', 'F']            → No loop
    
    Args:
        history: List of action names.
        window_size: Size of the sliding window to analyze.
        min_repetitions: Minimum consecutive repetitions to flag as loop.
        
    Returns:
        LoopDetectionResult with detection status.
    """
    if len(history) < window_size:
        return LoopDetectionResult(loop_detected=False)

    window = history[-window_size:]
    half = len(window) // 2

    for pattern_len in range(1, half + 1):
        pattern = window[:pattern_len]

        # Count consecutive repetitions of the pattern
        reps = 0
        for i in range(0, len(window), pattern_len):
            chunk = window[i:i + pattern_len]
            if chunk == pattern:
                reps += 1
            else:
                break

        if reps >= min_repetitions:
            return LoopDetectionResult(
                loop_detected=True,
                pattern=pattern,
                repetitions=reps,
                window_size=window_size,
            )

    return LoopDetectionResult(
        loop_detected=False,
        window_size=window_size,
    )

File: sdk-ts/src/loop-detector.ts
/**
 * SteerPlane SDK — Loop Detection Engine
 *
 * Sliding window pattern detector that identifies repeating action sequences.
 * Detects patterns like [A, B, A, B, A, B] or [A, A, A, A, A].
 */

export interface LoopDetectionResult {
  loopDetected: boolean;
  pattern: string[];
  repetitions: number;
  windowSize: number;
}

/**
 * Detect repeating action sequences using a sliding window.
 *
 * The check is anchored at the *end* of the window (matching the Python SDK), so
 * a loop that starts partway through the window (e.g. [X, A, B, A, B, A, B]) is
 * still detected.
 *
 * Algorithm:
 * 1. Take the last `windowSize` actions.
 * 2. For each candidate pattern length (1 .. window / minRepetitions):
 *    - Take the trailing `patternLen` actions as the candidate pattern.
 *    - Walk backwards counting consecutive repetitions from the end.
 *    - If repetitions >= the required count, it's a loop (smallest unit wins).
 *
 * Single-action loops (pattern length 1) require >= 3 consecutive repetitions —
 * a benign double-call should not terminate a run. Multi-step patterns use
 * `minRepetitions`.
 *
 * @example
 * detectLoop(['search', 'search', 'search'], 3)  → loop (pattern: ['search'])
 * detectLoop(['A', 'B', 'A', 'B', 'A', 'B'], 6)  → loop (pattern: ['A', 'B'])
 */
export function detectLoop(
  history: string[],
  windowSize: number = 8,
  minRepetitions: number = 2
): LoopDetectionResult {
  if (history.length < windowSize) {
    return { loopDetected: false, pattern: [], repetitions: 0, windowSize };
  }

  const window = history.slice(-windowSize);
  const n = window.length;
  const maxPatternLen = Math.max(1, Math.floor(n / minRepetitions));

  const equal = (a: string[], b: string[]): boolean =>
    a.length === b.length && a.every((v, idx) => v === b[idx]);

  for (let patternLen = 1; patternLen <= maxPatternLen; patternLen++) {
    const pattern = window.slice(n - patternLen, n);

    let reps = 0;
    let idx = n;
    while (idx - patternLen >= 0 && equal(window.slice(idx - patternLen, idx), pattern)) {
      reps++;
      idx -= patternLen;
    }

    // Single-action loops need a higher floor (>=3); multi-step use minRepetitions.
    const required = patternLen > 1 ? minRepetitions : Math.max(minRepetitions, 3);
    if (reps >= required) {
      return { loopDetected: true, pattern, repetitions: reps, windowSize };
    }
  }

  return { loopDetected: false, pattern: [], repetitions: 0, windowSize };
}

export class LoopDetector {
  private readonly windowSize: number;
  private readonly minRepetitions: number;
  private actionHistory: string[] = [];

  constructor(windowSize: number = 8, minRepetitions: number = 2) {
    this.windowSize = windowSize;
    this.minRepetitions = minRepetitions;
  }

  /** Record an action and check for loops. */
  recordAction(action: string): LoopDetectionResult {
    this.actionHistory.push(action);
    return this.check();
  }

  /** Check the current action history for loops. */
  check(): LoopDetectionResult {
    if (this.actionHistory.length < this.windowSize) {
      return {
        loopDetected: false,
        pattern: [],
        repetitions: 0,
        windowSize: this.windowSize,
      };
    }
    return detectLoop(
      this.actionHistory,
      this.windowSize,
      this.minRepetitions
    );
  }

  /** Clear the action history. */
  reset(): void {
    this.actionHistory = [];
  }
}
